src/native/corehost/hostmisc/utils.h

```diff
@@ -19,10 +19,42 @@
 #define DOTNET_CORE_DOWNLOAD_URL _X("https://aka.ms/dotnet-download")
 #define DOTNET_CORE_APPLAUNCH_URL _X("https://aka.ms/dotnet-core-applaunch")
 
+#define DOTNET_INFO_URL _X("https://aka.ms/dotnet/runtimes-sdk-info")
+#define DOTNET_APP_LAUNCH_FAILED_URL _X("https://aka.ms/dotnet/app-launch-failed")
+#define DOTNET_SDK_NOT_FOUND_URL _X("https://aka.ms/dotnet/sdk-not-found")
+
+// This message is defined here for consistency between errors on the command line and GUI (Windows apphost).
+#define INSTALL_OR_UPDATE_NET_ERROR_MESSAGE _X("You must install or update .NET to run this application.")
+
 #define RUNTIME_STORE_DIRECTORY_NAME _X("store")
 
 bool ends_with(const pal::string_t& value, const pal::string_t& suffix, bool match_case);
 bool starts_with(const pal::string_t& value, const pal::string_t& prefix, bool match_case);
+
+namespace utils
+{
+    template<size_t L>
+    inline constexpr size_t strlen(const pal::char_t(&)[L])
```

Comment:
Newbie question: What is this syntax? Can somebody point me somewhere to read up on it? (I tried searching, but without knowing the terminology... no luck).

Reply:
This is just normal C++ template magic. It has been available for as long as I can recall. The idea is that a statically defined array length can be encoding in a template. This is very different from C where array arguments don't technically have any expressive value. The high-level documentation for this is defined as [non-type template parameters](https://en.cppreference.com/w/cpp/language/template_parameters#Non-type_template_parameter). However, that may not help elucidate what is actually going on.

An array constant in C++ is defined as `<TYPE> &[N]`, where `N` is determined at compile-time. Since this is fully expressed in C++, one can use templates to decompose the type using non-type template parameters&mdash;see above. Starting from arrays, we know that C++ string literals are simply character type arrays so it follows that it can apply there as well. A string literal in C++ is typed as `<CHAR_TYPE> const &[N]`, where `N` is a compile-time constant. Now recall that a string is defined to have a `nullptr` as the last element so `N` would include that, hence the `-1`.

A generalized template for computing the size of any statically sized array is below, but note for string literals this isn't what most people would assume, since `strlen`/`wcslen` don't include the `nullptr`.

```c++
template <typename T, size_t L>
constexpr size_t array_len(T const (&)[L])
{
    return L;
}
```

Reply:
Thanks a lot for the explanation - to me it's both beautiful and horrible - not sure which one to pick 😇 .